core/server/common/src/main/java/alluxio/master/metastore/BlockStore.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.metastore;
+
+import alluxio.master.metastore.BlockStore.Block;
+import alluxio.proto.meta.Block.BlockLocation;
+import alluxio.proto.meta.Block.BlockMeta;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * The block store keeps track of block sizes and block locations.
+ *
+ * BlockStores require external synchronization for operations on the same block id. Operations on
+ * different block ids can be performed concurrently.
+ */
+@NotThreadSafe
+public interface BlockStore extends Iterable<Block> {
+  /**
+   * @param id a block id
+   * @return the block's metadata, or none if the block does not exist
+   */
+  Optional<BlockMeta> getBlock(long id);
+
+  /**
+   * Adds block metadata to the block store. If the block already exists, its metadata will be
+   * updated to the given metadata.
+   *
+   * @param id the block id
+   * @param meta the block metadata
+   */
+  void putBlock(long id, BlockMeta meta);
+
+  /**
+   * @param id a block id to remove
+   */
+  void removeBlock(long id);
+
+  /**
+   * Removes all metadata from the block store.
+   */
+  void clear();
+
+  /**
+   * @param id a block id
+   * @return the locations of the block
+   */
+  List<BlockLocation> getLocations(long id);
+
+  /**
+   * Adds a new block location.
+   *
+   * @param id       a block id
+   * @param location a block location
+   */
+  void addLocation(long id, BlockLocation location);
```

Comment:
what if block exists already?

Reply:
nothing happens if the location already exists, updated the javadoc